File: frontend/src/api/agents.ts
// Copyright (c) 2026, Invergent SA, developed by Flavius Burca
// SPDX-License-Identifier: AGPL-3.0-only
//
import { authFetch } from "@/api/auth";

// ── Types ──────────────────────────────────────────────────────────

export interface AgentResponse {
  id: string;
  project_id: string;
  project_name: string;
  name: string;
  harness: string;
  display_name: string;
  description: string;
  version: string;
  model_id: string;
  model_name: string;
  status: string;
  replicas: Record<string, number> | null;
  image: string;
  endpoint: string;
  system_prompt: string;
  env_vars: Record<string, string> | null;
  resources: Record<string, string> | null;
  created_by_id: string;
  created_by_username: string;
  hub_ref: string | null;
  created_at: string | null;
}

export interface AgentListResponse {
  agents: AgentResponse[];
  total: number;
}

export interface AgentCreateRequest {
  name: string;
  harness: string;
  display_name: string;
  description?: string;
  version?: string;
  model_id: string;
  status?: string;
  replicas?: Record<string, number>;
  image?: string;
  endpoint?: string;
  system_prompt?: string;
  env_vars?: Record<string, string>;
  resources?: Record<string, string>;
}

export interface AgentUpdateRequest {
  name?: string;
  harness?: string;
  display_name?: string;
  description?: string;
  version?: string;
  model_id?: string;
  status?: string;
  replicas?: Record<string, number>;
  image?: string;
  endpoint?: string;
  system_prompt?: string;
  env_vars?: Record<string, string>;
  resources?: Record<string, string>;
}

// ── API calls ──────────────────────────────────────────────────────

export async function listAgents(
  projectId?: string,
  status?: string,
  harness?: string,
  limit?: number,
): Promise<AgentListResponse> {
  const params = new URLSearchParams();
  if (projectId) params.append("project_id", projectId);
  if (status) params.append("status", status);
  if (harness) params.append("harness", harness);
  if (limit != null) params.append("limit", String(limit));
  const response = await authFetch(`/api/agents/agents?${params}`);
  if (!response.ok) {
    const err = (await response.json().catch(() => null)) as { detail?: string } | null;
    throw new Error(err?.detail ?? "Failed to fetch agents");
  }
  return (await response.json()) as AgentListResponse;
}

export async function getAgent(agentId: string): Promise<AgentResponse> {
  const response = await authFetch(`/api/agents/agents/${agentId}`);
  if (!response.ok) {
    const err = (await response.json().catch(() => null)) as { detail?: string } | null;
    throw new Error(err?.detail ?? "Failed to fetch agent");
  }
  return (await response.json()) as AgentResponse;
}

export async function createAgent(
  projectId: string,
  body: AgentCreateRequest,
): Promise<AgentResponse> {
  const params = new URLSearchParams({ project_id: projectId });
  const response = await authFetch(`/api/agents/agents?${params}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!response.ok) {
    const err = (await response.json().catch(() => null)) as { detail?: string } | null;
    throw new Error(err?.detail ?? "Failed to create agent");
  }
  return (await response.json()) as AgentResponse;
}

export async function updateAgent(
  agentId: string,
  body: AgentUpdateRequest,
): Promise<AgentResponse> {
  const response = await authFetch(`/api/agents/agents/${agentId}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!response.ok) {
    const err = (await response.json().catch(() => null)) as { detail?: string } | null;
    throw new Error(err?.detail ?? "Failed to update agent");
  }
  return (await response.json()) as AgentResponse;
}

export async function deleteAgent(agentId: string): Promise<void> {
  const response = await authFetch(`/api/agents/agents/${agentId}`, {
    method: "DELETE",
  });
  if (!response.ok) {
    const err = (await response.json().catch(() => null)) as { detail?: string } | null;
    throw new Error(err?.detail ?? "Failed to delete agent");
  }
}

File: surogate/server/routes/agents.py
"""Agent CRUD API routes."""

from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from surogate.core.db.engine import get_session
from surogate.core.db.models.operate import AgentStatus
from surogate.core.db.repository import agents as agent_repo
from surogate.core.db.repository import user as user_repo
from surogate.server.auth.authentication import get_current_subject
from surogate.server.models.agent import (
    AgentCreateRequest,
    AgentListResponse,
    AgentResponse,
    AgentUpdateRequest,
)

router = APIRouter()

# ── helpers ──────────────────────────────────────────────────────────


def _agent_to_response(agent) -> AgentResponse:
    return AgentResponse(
        id=agent.id,
        project_id=agent.project_id,
        project_name=agent.project.name if agent.project else "",
        name=agent.name,
        harness=agent.harness,
        display_name=agent.display_name,
        description=agent.description or "",
        version=agent.version,
        model_id=agent.model_id,
        model_name=agent.model.name if agent.model else "",
        status=agent.status.value if hasattr(agent.status, "value") else agent.status,
        replicas=agent.replicas,
        image=agent.image or "",
        endpoint=agent.endpoint or "",
        system_prompt=agent.system_prompt or "",
        env_vars=agent.env_vars,
        resources=agent.resources,
        created_by_id=agent.created_by_id,
        created_by_username=agent.created_by.username if agent.created_by else "",
        hub_ref=agent.hub_ref,
        created_at=str(agent.created_at) if agent.created_at else None,
    )


# ── Agent CRUD ───────────────────────────────────────────────────────


@router.get("/agents", response_model=AgentListResponse)
async def list_agents(
    current_subject: str = Depends(get_current_subject),
    session: AsyncSession = Depends(get_session),
    project_id: Optional[str] = Query(None),
    status: Optional[str] = Query(None),
    harness: Optional[str] = Query(None),
    limit: int = Query(100, le=500),
):
    agents = await agent_repo.list_agents(
        session,
        project_id=project_id,
        status=status,
        harness=harness,
        limit=limit,
    )
    return AgentListResponse(
        agents=[_agent_to_response(a) for a in agents],
        total=len(agents),
    )


@router.get("/agents/{agent_id}", response_model=AgentResponse)
async def get_agent(
    agent_id: str,
    current_subject: str = Depends(get_current_subject),
    session: AsyncSession = Depends(get_session),
):
    agent = await agent_repo.get_agent(session, agent_id)
    if agent is None:
        raise HTTPException(status_code=404, detail="Agent not found")
    return _agent_to_response(agent)


@router.post("/agents", response_model=AgentResponse, status_code=201)
async def create_agent(
    body: AgentCreateRequest,
    current_subject: str = Depends(get_current_subject),
    session: AsyncSession = Depends(get_session),
    project_id: str = Query(...),
):
    user = await user_repo.get_user_by_username(session, current_subject)
    if user is None:
        raise HTTPException(status_code=401, detail="User not found")

    try:
        agent_status = AgentStatus(body.status)
    except ValueError:
        raise HTTPException(status_code=400, detail=f"Invalid status: {body.status}")

    agent = await agent_repo.create_agent(
        session,
        project_id=project_id,
        name=body.name,
        harness=body.harness,
        display_name=body.display_name,
        description=body.description,
        version=body.version,
        model_id=body.model_id,
        status=agent_status,
        replicas=body.replicas,
        image=body.image,
        endpoint=body.endpoint,
        system_prompt=body.system_prompt,
        env_vars=body.env_vars,
        resources=body.resources,
        created_by_id=user.id,
    )
    return _agent_to_response(agent)


@router.patch("/agents/{agent_id}", response_model=AgentResponse)
async def update_agent(
    agent_id: str,
    body: AgentUpdateRequest,
    current_subject: str = Depends(get_current_subject),
    session: AsyncSession = Depends(get_session),
):
    existing = await agent_repo.get_agent(session, agent_id)
    if existing is None:
        raise HTTPException(status_code=404, detail="Agent not found")

    fields = body.model_dump(exclude_unset=True)
    if not fields:
        return _agent_to_response(existing)

    agent = await agent_repo.update_agent(session, agent_id, **fields)
    if agent is None:
        raise HTTPException(status_code=404, detail="Agent not found")
    return _agent_to_response(agent)


@router.delete("/agents/{agent_id}", status_code=204)
async def delete_agent(
    agent_id: str,
    current_subject: str = Depends(get_current_subject),
    session: AsyncSession = Depends(get_session),
):
    existing = await agent_repo.get_agent(session, agent_id)
    if existing is None:
        raise HTTPException(status_code=404, detail="Agent not found")

    await agent_repo.delete_agent(session, agent_id)
